language/move-stdlib/src/natives/debug.rs

```diff
@@ -60,29 +61,59 @@ fn native_print(
     #[cfg(feature = "testing")]
     {
         let ty = ty_args.pop().unwrap();
-
-        match is_string_type(_context, &ty, move_std_addr) {
-            true => {
-                let r = pop_arg!(args, StructRef);
-
-                let field_ref = r
-                    .borrow_field(0)? // borrows the 'bytes' field of the std::string::String struct
-                    .value_as::<VectorRef>()?;
-
-                let bytes = field_ref.as_bytes_ref();
-
-                // This is safe because we guarantee the bytes to be utf8.
-                let str = unsafe { std::str::from_utf8_unchecked(bytes.as_slice()) };
-
-                println!("[debug] {}", str);
-            }
-            false => {
-                let r = pop_arg!(args, Reference);
-                let mut buf = String::new();
-                print_reference(&mut buf, &r)?;
-                println!("[debug] {}", buf);
+        
+        // if ty is a struct type:
+        // get type layout in VM format (raw BCS, no field names) 
+        let old_type_layout = context.type_to_type_layout(&ty)?.unwrap();
+
+        println!("Type: {:?}", &ty);
+        println!("Value: {:?}", &args[0]);
+        println!("MoveTypeLayout: {:?}", old_type_layout);
+
+        // For some reason this line fails when calling `debug::print` on the `TestStruct` in debug.move
```

Comment:
I got this to work on your tests with the following tweaks:

```
  let ty = ty_args.pop().unwrap();
        // get type layout in VM format (raw BCS, no field names)
        let old_layout = context.type_to_type_layout(&ty)?.unwrap();
        match &old_layout {
            MoveTypeLayout::Struct(old_struct_layout) => {
                let arg = pop_arg!(args, StructRef);
                let arg_val = arg.read_ref()?;
                // get BCS bytes for the struct
                let blob = arg_val.simple_serialize(&old_layout).unwrap();
                let move_struct =
                    MoveStruct::simple_deserialize(&blob, &old_struct_layout).unwrap();

                // get type layout in decorated format (with field names)
                let new_type_layout = context.type_to_fully_annotated_layout(&ty)?.unwrap();
                let new_struct_layout = match &new_type_layout {
                    MoveTypeLayout::Struct(struct_layout) => struct_layout,
                    _ => panic!("ooppsie; this shouldn't happen."),
                };

                let decorated_struct = move_struct.decorate(new_struct_layout);
                println!("[debug] {}", decorated_struct);
```

Reply:
I then get:

```
INCLUDING DEPENDENCY MoveStdlib
BUILDING MoveNursery
Running Move unit tests
[ PASS    ] 0x1::errors_tests::errors_state
[ PASS    ] 0x1::guid_tests::test_basics
[ PASS    ] 0x1::roleTests::test_assign_failure
[ PASS    ] 0x1::aclTests::test_add_failure
[ PASS    ] 0x1::roleTests::test_revoke_failure
[ PASS    ] 0x1::guid_tests::test_delegation
[ PASS    ] 0x1::capability_tests::test_delegate_failure_after_revoke
[debug] 0x1::debug::TestStruct {number: 255u8, bytes: vector[192u8, 255u8, 238u8], name: 0x1::string::String {bytes: vector[72u8, 101u8, 108u8, 108u8, 111u8, 44u8, 32u8, 115u8, 97u8, 110u8, 101u8, 32u8, 77u8, 111u8, 118u8, 101u8, 32u8, 100u8, 101u8, 98u8, 117u8, 103u8, 103u8, 105u8, 110u8, 103u8, 33u8]}}
[ PASS    ] 0x1::debug::test_print_annotated_struct
[ PASS    ] 0x1::roleTests::test_success
[ PASS    ] 0x1::guid_tests::test_id
[ PASS    ] 0x1::compareTests::equality_of_simple_types
[ PASS    ] 0x1::aclTests::test_remove_failure
[ PASS    ] 0x1::event_tests::test_event_128
[ PASS    ] 0x1::capability_tests::test_delegate_success
[debug] (&) [72, 101, 108, 108, 111, 44, 32, 115, 97, 110, 101, 32, 77, 111, 118, 101, 32, 100, 101, 98, 117, 103, 103, 105, 110, 103, 33]
[debug] 0x1::string::String {bytes: vector[72u8, 101u8, 108u8, 108u8, 111u8, 44u8, 32u8, 115u8, 97u8, 110u8, 101u8, 32u8, 77u8, 111u8, 118u8, 101u8, 32u8, 100u8, 101u8, 98u8, 117u8, 103u8, 103u8, 105u8, 110u8, 103u8, 33u8]}
```

which is ugly, but does have field names!

Reply:
I think `struct DebugStruct(MoveStruct)` with a custom `Display` impl should do the trick for the rest.

Reply:
Thank you!

The following struct:
```
        let obj = TestStruct {
            number: 255u8,
            bytes: x"c0ffee",
            name: string::utf8(b"Hello, sane Move debugging!"),
        };
```

...now prints as:
```
[annotated-debug] 0x1::debug::TestStruct {number: 255u8, bytes: vector[192u8, 255u8, 238u8], name: 0x1::string::String {bytes: vector[72u8, 101u8, 108u8, 108u8, 111u8, 44u8, 32u8, 115u8, 97u8, 110u8, 101u8, 32u8, 77u8, 111u8, 118u8, 101u8, 32u8, 100u8, 101u8, 98u8, 117u8, 103u8, 103u8, 105u8, 110u8, 103u8, 33u8]}}
```

This seems to be enough for pretty-print and, as a result, could be enough for domain-separated signatures.

I could change this into a PR that adds a new `pretty_print` function in `std::debug` if you all think it's a good idea @sblackshear, @wrwg, @vgao1996 et al.

(This would also be a good way for me to familiarize myself with these "decorated" `MoveStruct`'s.)